### PR TITLE
[benchmark] Fix broken armv8 builds by not inferring 32-bit build option

### PR DESCRIPTION
**Pull request by ralfholly**
Specify library name and version:  **lib/1.0**

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

# Reason for change
Currently, all builds for armv8 (like Raspberry PI 4, NVIDIA Jetson) fail because of a check that makes no sense for armv8 platforms. The CMake option BENCHMARK_BUILD_32_BITS is automatically set to ON if the settings.arch string doesn't contain the number "64". This can never be the case if settings.arch == "armv8" (which is a 64-bit platform). Hence, on armv8 platforms, BENCHMARK_BUILD_32_BITS is always on, which during compilation leads to the use of the '-m32' compiler switch. This switch is not supported by gcc, so the building with gcc on armv8 is not possible. This is a BIG BIG BIG problem!

# Change
After the fix, the CMake BENCHMARK_BUILD_32_BITS option is not derived from the value of the settings.arch string. If a user wants a 32-bit build, they have to manually set the option "build_32_bits" to set the corresponding CMake option "BENCHMARK_BUILD_32_BITS".

### CODE REVIEW

**Comment by ralfholly** [comment]
@Croydon We already had a discussion about this on the benchmark repo.

File: recipes/benchmark/all/conanfile.py
Comment (mathbunnyru) [inline recipes/benchmark/all/conanfile.py:24]:
```suggestion
    default_options = {
        "shared": False,
        "fPIC": True,
        "enable_lto": False,
        "enable_exceptions": True,
        "build_32_bits": False
    }
```

File: recipes/benchmark/all/conanfile.py
Comment (mathbunnyru) [inline recipes/benchmark/all/conanfile.py:57]:
```suggestion
```

**Review by mathbunnyru — CHANGES_REQUESTED** [review]
Maybe a better default option would be something like this?

```
        arch64 = ['x86_64', 'sparcv9', 'ppc64', 'ppc64le', 'armv8', 'armv8.3', 'mips64']
        bits = "64" if self.settings.arch in arch64 else "32"
```

**Comment by ralfholly** [comment]
> Maybe a better default option would be something like this?
> 
> ```
>         arch64 = ['x86_64', 'sparcv9', 'ppc64', 'ppc64le', 'armv8', 'armv8.3', 'mips64']
>         bits = "64" if self.settings.arch in arch64 else "32"
> ```

IMHO, the underlying CMakeLists.txt uses an explicit option to enable 32-bit builds and so should the conanfile. Your solution solves the immediate problem, but the arch64 list needs maintenance.

**Comment by mathbunnyru** [comment]
> > Maybe a better default option would be something like this?
> > ```
> >         arch64 = ['x86_64', 'sparcv9', 'ppc64', 'ppc64le', 'armv8', 'armv8.3', 'mips64']
> >         bits = "64" if self.settings.arch in arch64 else "32"
> > ```
> 
> IMHO, the underlying CMakeLists.txt uses an explicit option to enable 32-bit builds and so should the conanfile. Your solution solves the immediate problem, but the arch64 list needs maintenance.

Sorry, I chose a very bad explanation.

What I meant is to still add the explicit version (to follow the upstream CMake), like you did, but for the cases, where we know for sure we need 32 bits, we might want to change the default value to True.
To do that, it's possible to add `config_options` method, which will set `build_32_bits` to True, when we are sure that we want 32bits version (for example, "x86" arch).
This will allow building this recipe with default settings under "x86" architecture.

**Comment by ralfholly** [comment]
> > > Maybe a better default option would be something like this?
> > > ```
> > >         arch64 = ['x86_64', 'sparcv9', 'ppc64', 'ppc64le', 'armv8', 'armv8.3', 'mips64']
> > >         bits = "64" if self.settings.arch in arch64 else "32"
> > > ```
> > 
> > 
> > IMHO, the underlying CMakeLists.txt uses an explicit option to enable 32-bit builds and so should the conanfile. Your solution solves the immediate problem, but the arch64 list needs maintenance.
> 
> Sorry, I chose a very bad explanation.
> 
> What I meant is to still add the explicit version (to follow the upstream CMake), like you did, but for the cases, where we know for sure we need 32 bits, we might want to change the default value to True.
> To do that, it's possible to add `config_options` method, which will set `build_32_bits` to True, when we are sure that we want 32bits version (for example, "x86" arch).
> This will allow building this recipe with default settings under "x86" architecture.

Thanks for your feedback. Honestly, I have no strong opinion on that. To me, and possibly many, many other developers working on armv8-based projects, the most important thing is that we get rid of this nasty build-breaker as soon as possible.

**Comment by ralfholly** [comment]
This is related to newly created issue #4677

**Comment by mathbunnyru** [comment]
@ralfholly to get this merged, you will still have to apply to EAP (only after this PRs are tested in the CI).

So, please, ask to be included in EAP here:
https://github.com/conan-io/conan-center-index/issues/4

And, please, apply my review suggestions - this will not slow down the merging of this PR.

**Comment by SpaceIm** [comment]
I disagree with this modification, while agree that something is wrong in previous revision of this recipe.
I would not add an option, just turn off this CMakeList option unconditionally to not enter in this branch https://github.com/google/benchmark/blob/5c43112eb3d70cec10784d8d1f4373bc7397836b/CMakeLists.txt#L97.
flags related to arch should come from conan profile. We do not have to mimic all upstream options, specially when they violate conan paradigm.

**Comment by ralfholly** [comment]
> I disagree with this modification, while agree that something is wrong in previous revision of this recipe.
> I would not add an option, just turn off this CMakeList option unconditionally to not enter in this branch https://github.com/google/benchmark/blob/5c43112eb3d70cec10784d8d1f4373bc7397836b/CMakeLists.txt#L97.
> flags related to arch should come from conan profile. We do not have to mimic all upstream options, specially when they violate conan paradigm.

Makes sense to me! How do we go about this? Shall I close this PR and create a new one?

**Comment by mathbunnyru** [comment]
Please, use this PR (do not create new one), it will be easier.

**Review by mathbunnyru — APPROVED** [review]

**Review by SpaceIm — APPROVED** [review]

**Review by prince-chrismc — APPROVED** [review]
Murdering 32-bit support? Sign me up!

**Comment by ralfholly** [comment]
Hmm, I do have three approvals, but it looks like the approvers must also have write access.

**Comment by prince-chrismc** [comment]
> Hmm, I do have three approvals, but it looks like the approvers must also have write access.

I am on my phone but in the docs we have "review process" that has all the details about how it works. It should not be long (theres a lot of prs recently)

**Comment by mathbunnyru** [comment]
https://github.com/conan-io/conan-center-index/blob/a433dd4cc628384bf6500d73e6f860e35affa49d/docs/review_process.md#rule-of-3-reviews

> but it couldn't be just 3 community reviews

right now there are only 3 community reviewers.

**Review by uilianries — APPROVED** [review]

**Comment by ralfholly** [comment]
Thanks to everyone helping me getting this done!

**Pull request merged**
